### ISSUE TITLE
an approximate solution to compute mutual information

### ISSUE THREAD

**Issue by argman** [voxelmorph/voxelmorph#31]
@adalca @balakg 
Hi，I find an approximate solution to compute mutual information in tensorflow, but I cannot test its rightness as I donot have enough data to train the model, can anyone help ? code is here:
```
def nmi_gaussian(R, T, win=20, eps=1e-5):
    '''
        Parzen window approximation of mutual information
    Params:
        R : Reference(Fixed) Image, shape should be N * H * W * Z * 1
        T: Test (Moving) Image, shape should be the same as R
        win: number of bins used in histogram counting
    '''
    N, H, W, Z, C = R.shape
    assert C == 1, 'image should be only one channel'
    im_size = N.value * H.value * W.value * Z.value

    R_min = tf.reduce_min(R, keep_dims=False)
    R_max = tf.reduce_max(R, keep_dims=False)
    T_min = tf.reduce_min(T, keep_dims=False)
    T_max = tf.reduce_max(T, keep_dims=False)

    R_bin_size = (R_max - R_min) / win
    T_bin_size = (T_max - T_min) / win

    # compute bins
    R_bin_window = tf.range(R_min + 0.5 * R_bin_size, R_min + 0.5 * R_bin_size + R_bin_size * win - eps, delta=R_bin_size)
    T_bin_window = tf.range(T_min + 0.5 * T_bin_size, T_min + 0.5 * T_bin_size + T_bin_size * win - eps, delta=T_bin_size)

    R_mesh = tf.tile(tf.reshape(R_bin_window, (-1, 1)), multiples=[1, win])
    T_mesh = tf.tile(tf.reshape(T_bin_window, (1, -1)), multiples=[win, 1])
    R_T_mesh = tf.concat([tf.reshape(R_mesh, (-1, 1)), tf.reshape(T_mesh, (-1, 1))], axis=-1)
    R_T_mesh = R_T_mesh[tf.newaxis, tf.newaxis, tf.newaxis, :, :]

    p_l_k = 1/(np.sqrt(2 * np.pi)) * tf.exp(-0.5 * (tf.square((R - R_T_mesh[..., 0])/R_bin_size) + tf.square((T - R_T_mesh[..., 1])/T_bin_size)))
    
    p_l_k = tf.reduce_sum(p_l_k, axis=(0, 1, 2, 3)) / im_size
    p_l_k = p_l_k / tf.reduce_sum(p_l_k)
    p_l_k = tf.reshape(p_l_k, (win, win))
    p_l = tf.reduce_sum(p_l_k, axis=0)
    p_k = tf.reduce_sum(p_l_k, axis=1)

    pl_pk = p_l[:, tf.newaxis] * p_k[tf.newaxis, :]

    mi = p_l_k * tf.log(p_l_k / pl_pk)

    mi = tf.where(tf.is_finite(mi), mi, tf.zeros_like(mi))
    mi = -tf.reduce_sum(mi)
    return mi
```

the idea is using Parzen Window estimation of MutualInformation, as of which used in MattesMutualInformation of ITK, but replace the B-Spline window function with a Gaussian window function

_Originally posted by @argman in https://github.com/voxelmorph/voxelmorph/issues/25#issuecomment-477512229_

**Comment by adalca** [comment]
hi @argman , C+P from my answer just now on the other issue:

> Yes, thank you for the code -- we're just in the middle of deadline season (MICCAI is on tuesday) and most of us are working on that. We'll be happy to get back to you after this!

**Comment by nazib** [comment]
Hi @adalca 
                  I created a patch-based version of your model. Therefore, we feed randomly selected small patches of size 64x64x64 into the voxelmorph network and train it same as before. During test, we collect patches with 50% overlap and feed them to a trained model. The model gives a flow field which is than applied to the moving image patch. After coillecting all te patches we combine them with sigmoid like weighting function. In this way we get a full resolution image registered. But the problem is the flow field generated from the patches are not smooth and there is always some boundary effect due to discountinous flow in the full resolution image. Did you ever tried your method in patch based registration? If so how do you mange to  get patch by patch smooth (continous) flow field.
Sorry to comment here since my question is totally unrelated to this issue.
Kind Regards
Nazib

**Comment by adalca** [comment]
hi @nazib, please start a new issue, or was there a reason you didn't?

**Comment by nazib** [comment]
Ok

**Comment by nazib** [comment]
Hi @argman, is this implementation of MI correct? I tried to train VM with this loss and its not learning at all.
Hi @adalca, have you ever tried to train VM with MI loss? If yes, what is the result. It would be helpful for us if you share experience on MI loss with us.
Kind regards

**Comment by farhanone** [comment]
Hello @nazib, I have tested the code by @argman, but it requires much GPU memory to compute MI for bins > 10. It shows less value of MI when using same tensor as input  `nmi_gaussian(R, R, win=10, eps=1e-5)`result = -0.19 than  `nmi_gaussian(R, T, win=10, eps=1e-5)` with result = -0.42.

**Comment by adalca** [comment]
@nazib @farhanone we have a implementation we've been sending around for NMI, feel free to email me and I will send it to you. We'll have it in the official doc once everything is reorganized.

@nazib yes we've tried the NMI loss for voxelmorph, and it worked well for registering different modalities within MR.

**Comment by farhanone** [comment]
@adalca Thanks. I have already sent you an email for the code...

**Comment by adalca** [comment]
@farhanone you mean before or just recently? Anyway, if I missed it, just re-email please. It could have slipped through the track.

**Comment by farhanone** [comment]
@adalca I just sent you the email again. If there is some problem getting the email then, please send me at 
farhan.one@hotmail.com

**Comment by adalca** [comment]
@farhanone done.

**Comment by farhanone** [comment]
@adalca many thanks

**Comment by nazib** [comment]
Hi @adalca 
                 Thanks a lot I will send you email now. 
Hi @farhanone 
                  Thanks for sharing your experience on MI loss. I also tried with this MI code and It was not good. Lets see how Adrian solved this issue. I will try to share my experience after retraining my model.

**Comment by nazib** [comment]
Hi @adalca,
                  Could you please check your email.

**Comment by adalca** [comment]
@nazib done

**Comment by junyuchen245** [comment]
Hi @adalca 
Thank you so much for sharing the VoxelMorph source code. I think this is great work!
Could you please share the NMI implementation with me? I have been searching for it for weeks.
My email address is jchen245@jhu.edu.
Many thanks!

**Comment by adalca** [comment]
@junyuchen245 done!

**Comment by a-kudoh** [comment]
Hi @adalca, Thanks for sharing your great work!
I am also suffering in NMI implementation.  Could you please share the implementation at 
akirdo@gmail.com
Thank you!

**Comment by norris9410** [comment]
Hi @adalca, thanks for making this open-source, I am wondering if you could also send me your implementation of NMI whenever you are free? It will save me lot of time.
ncncnorris@gmail.com
Thanks in advance

**Comment by adalca** [comment]
done. I'll try to add this to the `losses.py` since this is not a sustainable model :)

**Comment by yangwyou** [comment]
Hi @adalca ,thanks for sharing the VM, I am training it with my dataset. I am wondering if you could send me your implementation of NMI if possible? I want to try it. My email is wenyouyang1029@gmail.com.  Thank you!

**Comment by oooo1114** [comment]
Hi, @adalca , thanks for sharing your work. I am working on cross-modality registration, could you send me the MI loss when you are free, many thanks, my email address is 
lvyuanyuanoo@gmail.com
thanks again.

**Comment by soanduong** [comment]
Hi @adalca, thanks for publishing your work.

Would you mind to send the MI loss? My email is soandm@gmail.com.

Thanks again.

**Comment by adalca** [comment]
@soanduong and everyone,

I've added an experimental NMI loss to `losses.py`, please take a look.

**Comment by soanduong** [comment]
@adalca thank you very much for the sharing.

**Comment by lucasestini95** [comment]
Hi @adalca , thanks a lot for your work and for making it open source.

Could you please send the MI loss? My email is luca171295@gmail.com

Thanks again.

**Comment by adalca** [comment]
@lucasestini95 the mutual information has been released for a while now!

We have one implemented in voxelmorph:
https://github.com/voxelmorph/voxelmorph/blob/dev/voxelmorph/tf/losses.py

We also have a newer, cleaner one implemented in neurite which will eventually take over:
https://github.com/adalca/neurite/blob/dev/neurite/tf/losses.py
